# Project: Stock Market

This data is taken from https://www.kaggle.com/cnic92/200-financial-indicators-of-us-stocks-20142018

## Context
The algorithmic trading space is buzzing with new strategies. Companies have spent billions in infrastructures and R&D to be able to jump ahead of the competition and beat the market. Still, it is well acknowledged that the buy & hold strategy is able to outperform many of the algorithmic strategies, especially in the long-run. However, finding value in stocks is an art that very few mastered, can a computer do that?

## Content
This Data repo contains the following datasets (in .csv format):

- 2014_Financial_Data.csv
- 2015_Financial_Data.csv
- 2016_Financial_Data.csv
- 2017_Financial_Data.csv
- 2018_Financial_Data.csv

Each dataset contains 200+ financial indicators, that are commonly found in the 10-K filings each publicly traded company releases yearly, for a plethora of US stocks (on average, 4k stocks are listed in each dataset). I built this dataset leveraging Financial Modeling Prep API and pandas_datareader.

## Important remarks regarding the datasets:

Some financial indicator values are missing (nan cells), so the user can select the best technique to clean each dataset (dropna, fillna, etc.).

There are outliers, meaning extreme values that are probably caused by mistypings. Also in this case, the user can choose how to clean each dataset (have a look at the 1% - 99% percentile values).

The third-to-last column, Sector, lists the sector of each stock. Indeed, in the US stock market each company is part of a sector that classifies it in a macro-area. Since all the sectors have been collected (Basic Materials, Communication Services, Consumer Cyclical, Consumer Defensive, Energy, Financial Services, Healthcare, Industrial, Real Estate, Technology and Utilities), the user has the option to perform per-sector analyses and comparisons.

The second-to-last column, PRICE VAR [%], lists the percent price variation of each stock for the year. For example, if we consider the dataset 2015_Financial_Data.csv, we will have:

200+ financial indicators for the year 2015;
percent price variation for the year 2016 (meaning from the first trading day on Jan 2016 to the last trading day on Dec 2016).
The last column, class, lists a binary classification for each stock, where

for each stock, if the PRICE VAR [%] value is positive, class = 1. From a trading perspective, the 1 identifies those stocks that an hypothetical trader should BUY at the start of the year and sell at the end of the year for a profit.
for each stock, if the PRICE VAR [%] value is negative, class = 0. From a trading perspective, the 0 identifies those stocks that an hypothetical trader should NOT BUY, since their value will decrease, meaning a loss of capital.
The columns PRICE VAR [%] and class make possible to use the datasets for both classification and regression tasks:

If the user wishes to train a machine learning model so that it learns to classify those stocks that in buy-worthy and not buy-worthy, it is possible to get the targets from the class column;
If the user wishes to train a machine learning model so that it learns to predict the future value of a stock, it is possible to get the targets from the PRICE VAR [%] column.

The code to beautify the tables:

In [1]:
%%HTML
<style type="text/css">
table.dataframe td, table.dataframe th {
    border: 1px  black solid !important;
  color: black !important;
}
</style>

We have to import necessary libraries and datasets.

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [3]:
df_14 = pd.read_csv('2014_Financial_Data.csv')
df_15 = pd.read_csv('2015_Financial_Data.csv')
df_16 = pd.read_csv('2016_Financial_Data.csv')
df_17 = pd.read_csv('2017_Financial_Data.csv')
df_18 = pd.read_csv('2018_Financial_Data.csv')
df_14.head()

Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0         PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
1       VIPS  3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
2         KR  9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
3        RAD  2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
4        GIS  1.790960e+10          0.0076     1.153980e+10  6.369800e+09   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0  0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
1  1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
2  0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
3  0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
4  0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   

   Interest Expense  ...  Receivables growth  Inventory Growth  Asset Growth  \
0      7.090000e+08  ...             -0.0187           -0.0217        0.0359   
1      1.214869e+07  ...                 NaN               NaN           NaN   
2      4.430000e+08  ...              0.0618            0.0981        0.1886   
3      4.245910e+08  ...              0.0211           -0.0510       -0.0189   
4      3.024000e+08  ...              0.0257            0.0090        0.0215   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0316       0.1228              0.0000   
1                          NaN          NaN              1.6484   
2                       0.3268       0.2738              0.0000   
3                       0.1963      -0.0458              0.0000   
4                       0.0274       0.1025              0.0000   

   SG&A Expenses Growth              Sector  2015 PRICE VAR [%]  Class  
0               -0.1746  Consumer Defensive           -9.323276      0  
1                1.7313  Consumer Defensive          -25.512193      0  
2                0.0234  Consumer Defensive           33.118297      1  
3               -0.0060  Consumer Defensive            2.752291      1  
4               -0.0220  Consumer Defensive           12.897715      1  

[5 rows x 225 columns]

In [4]:
print("Shape of data for Year 2014 is {}".format(df_14.shape))
print("Shape of data for Year 2015 is {}".format(df_15.shape))
print("Shape of data for Year 2016 is {}".format(df_16.shape))
print("Shape of data for Year 2017 is {}".format(df_17.shape))
print("Shape of data for Year 2018 is {}".format(df_18.shape))

Shape of data for Year 2014 is (3808, 225)
Shape of data for Year 2015 is (4120, 225)
Shape of data for Year 2016 is (4797, 225)
Shape of data for Year 2017 is (4960, 225)
Shape of data for Year 2018 is (4392, 225)


In [5]:
df_18.head()

Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0      CMCSA  9.450700e+10          0.1115     0.000000e+00  9.450700e+10   
1        KMI  1.414400e+10          0.0320     7.288000e+09  6.856000e+09   
2       INTC  7.084800e+10          0.1289     2.711100e+10  4.373700e+10   
3         MU  3.039100e+10          0.4955     1.250000e+10  1.789100e+10   
4         GE  1.216150e+11          0.0285     9.546100e+10  2.615400e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0  0.000000e+00  6.482200e+10        7.549800e+10      1.900900e+10   
1  0.000000e+00  6.010000e+08        3.062000e+09      3.794000e+09   
2  1.354300e+10  6.750000e+09        2.042100e+10      2.331600e+10   
3  2.141000e+09  8.130000e+08        2.897000e+09      1.499400e+10   
4  0.000000e+00  1.811100e+10        4.071100e+10     -1.455700e+10   

   Interest Expense  ...  Receivables growth  Inventory Growth  Asset Growth  \
0      3.542000e+09  ...              0.2570            0.0000        0.3426   
1      1.917000e+09  ...              0.0345           -0.0920       -0.0024   
2     -1.260000e+08  ...              0.1989            0.0387        0.0382   
3      3.420000e+08  ...              0.4573            0.1511        0.2275   
4      5.059000e+09  ...             -0.2781           -0.2892       -0.1575   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0722       0.7309              0.0000   
1                       0.0076      -0.0137              0.0000   
2                       0.1014      -0.0169              0.0390   
3                       0.6395      -0.5841              0.1738   
4                      -0.4487      -0.2297              0.0000   

   SG&A Expenses Growth             Sector  2019 PRICE VAR [%]  Class  
0                0.1308  Consumer Cyclical           32.794573      1  
1               -0.1265             Energy           40.588068      1  
2               -0.0942         Technology           30.295514      1  
3                0.0942         Technology           64.213737      1  
4                0.0308        Industrials           44.757840      1  

[5 rows x 225 columns]

In [6]:
def info_nulls(table):
    #this function will show the number of null values and the object type.
    info = pd.DataFrame(table.dtypes).T.rename(index={0:'column type'})
    info = info.append(pd.DataFrame(table.isnull().sum()).T.rename(index={0:'number of null values'}))
    info = info.append(pd.DataFrame(table.isnull().sum()/table.shape[0]*100).T.rename(index={0:'percentage of null values'}))
    display(info)

In [7]:
info_nulls(df_14)

Unnamed: 0  Revenue Revenue Growth Cost of Revenue  \
column type                   object  float64        float64         float64   
number of null values              0       44            236              74   
percentage of null values          0  1.15546        6.19748         1.94328   

                          Gross Profit R&D Expenses SG&A Expense  \
column type                    float64      float64      float64   
number of null values               52          136           59   
percentage of null values      1.36555      3.57143      1.54937   

                          Operating Expenses Operating Income  \
column type                          float64          float64   
number of null values                     63               55   
percentage of null values            1.65441          1.44433   

                          Interest Expense  ... Receivables growth  \
column type                        float64  ...            float64   
number of null values                   63  ...                290   
percentage of null values          1.65441  ...            7.61555   

                          Inventory Growth Asset Growth  \
column type                        float64      float64   
number of null values                  290          290   
percentage of null values          7.61555      7.61555   

                          Book Value per Share Growth Debt Growth  \
column type                                   float64     float64   
number of null values                             369         302   
percentage of null values                     9.69013     7.93067   

                          R&D Expense Growth SG&A Expenses Growth  Sector  \
column type                          float64              float64  object   
number of null values                    247                  243       0   
percentage of null values            6.48634               6.3813       0   

                          2015 PRICE VAR [%]  Class  
column type                          float64  int64  
number of null values                      0      0  
percentage of null values                  0      0  

[3 rows x 225 columns]

In [8]:
info_nulls(df_15)

Unnamed: 0  Revenue Revenue Growth Cost of Revenue  \
column type                   object  float64        float64         float64   
number of null values              0       67            340             193   
percentage of null values          0  1.62621        8.25243         4.68447   

                          Gross Profit R&D Expenses SG&A Expense  \
column type                    float64      float64      float64   
number of null values               65          265          170   
percentage of null values      1.57767      6.43204      4.12621   

                          Operating Expenses Operating Income  \
column type                          float64          float64   
number of null values                    174               42   
percentage of null values             4.2233          1.01942   

                          Interest Expense  ... Receivables growth  \
column type                        float64  ...            float64   
number of null values                  175  ...                368   
percentage of null values          4.24757  ...            8.93204   

                          Inventory Growth Asset Growth  \
column type                        float64      float64   
number of null values                  369          369   
percentage of null values          8.95631      8.95631   

                          Book Value per Share Growth Debt Growth  \
column type                                   float64     float64   
number of null values                             458         399   
percentage of null values                     11.1165     9.68447   

                          R&D Expense Growth SG&A Expenses Growth  Sector  \
column type                          float64              float64  object   
number of null values                    351                  347       0   
percentage of null values            8.51942              8.42233       0   

                          2016 PRICE VAR [%]  Class  
column type                          float64  int64  
number of null values                      0      0  
percentage of null values                  0      0  

[3 rows x 225 columns]

In [9]:
info_nulls(df_16)

Unnamed: 0  Revenue Revenue Growth Cost of Revenue  \
column type                   object  float64        float64         float64   
number of null values              0      489            649             640   
percentage of null values          0  10.1939        13.5293         13.3417   

                          Gross Profit R&D Expenses SG&A Expense  \
column type                    float64      float64      float64   
number of null values              492          713          613   
percentage of null values      10.2564      14.8635      12.7788   

                          Operating Expenses Operating Income  \
column type                          float64          float64   
number of null values                    620              465   
percentage of null values            12.9247          9.69356   

                          Interest Expense  ... Receivables growth  \
column type                        float64  ...            float64   
number of null values                  620  ...                640   
percentage of null values          12.9247  ...            13.3417   

                          Inventory Growth Asset Growth  \
column type                        float64      float64   
number of null values                  725          722   
percentage of null values          15.1136      15.0511   

                          Book Value per Share Growth Debt Growth  \
column type                                   float64     float64   
number of null values                             873         796   
percentage of null values                     18.1989     16.5937   

                          R&D Expense Growth SG&A Expenses Growth  Sector  \
column type                          float64              float64  object   
number of null values                    772                  756       0   
percentage of null values            16.0934              15.7598       0   

                          2017 PRICE VAR [%]  Class  
column type                          float64  int64  
number of null values                      0      0  
percentage of null values                  0      0  

[3 rows x 225 columns]

In [10]:
info_nulls(df_17)

Unnamed: 0  Revenue Revenue Growth Cost of Revenue  \
column type                   object  float64        float64         float64   
number of null values              0      525            724             679   
percentage of null values          0  10.5847        14.5968         13.6895   

                          Gross Profit R&D Expenses SG&A Expense  \
column type                    float64      float64      float64   
number of null values              534          787          661   
percentage of null values      10.7661      15.8669      13.3266   

                          Operating Expenses Operating Income  \
column type                          float64          float64   
number of null values                    661              504   
percentage of null values            13.3266          10.1613   

                          Interest Expense  ... Receivables growth  \
column type                        float64  ...            float64   
number of null values                  677  ...                703   
percentage of null values          13.6492  ...            14.1734   

                          Inventory Growth Asset Growth  \
column type                        float64      float64   
number of null values                  811          792   
percentage of null values          16.3508      15.9677   

                          Book Value per Share Growth Debt Growth  \
column type                                   float64     float64   
number of null values                             893         854   
percentage of null values                      18.004     17.2177   

                          R&D Expense Growth SG&A Expenses Growth  Sector  \
column type                          float64              float64  object   
number of null values                    850                  828       0   
percentage of null values            17.1371              16.6935       0   

                          2018 PRICE VAR [%]  Class  
column type                          float64  int64  
number of null values                      0      0  
percentage of null values                  0      0  

[3 rows x 225 columns]

In [11]:
info_nulls(df_18)

Unnamed: 0  Revenue Revenue Growth Cost of Revenue  \
column type                   object  float64        float64         float64   
number of null values              0       46            139             185   
percentage of null values          0  1.04736        3.16485          4.2122   

                          Gross Profit R&D Expenses SG&A Expense  \
column type                    float64      float64      float64   
number of null values               64          237          166   
percentage of null values      1.45719      5.39617       3.7796   

                          Operating Expenses Operating Income  \
column type                          float64          float64   
number of null values                    184               35   
percentage of null values            4.18944         0.796903   

                          Interest Expense  ... Receivables growth  \
column type                        float64  ...            float64   
number of null values                  184  ...                124   
percentage of null values          4.18944  ...            2.82332   

                          Inventory Growth Asset Growth  \
column type                        float64      float64   
number of null values                  232          214   
percentage of null values          5.28233       4.8725   

                          Book Value per Share Growth Debt Growth  \
column type                                   float64     float64   
number of null values                             271         264   
percentage of null values                     6.17031     6.01093   

                          R&D Expense Growth SG&A Expenses Growth  Sector  \
column type                          float64              float64  object   
number of null values                    259                  248       0   
percentage of null values            5.89709              5.64663       0   

                          2019 PRICE VAR [%]  Class  
column type                          float64  int64  
number of null values                      0      0  
percentage of null values                  0      0  

[3 rows x 225 columns]

In [12]:
for col in df_14.columns:
    print(df_14[col].astype(bool))

0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: Unnamed: 0, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: Revenue, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: Revenue Growth, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: Cost of Revenue, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: Gross Profit, Length: 3808, dtype: bool
0       False
1        True
2       False
3       False
4

0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: freeCashFlowOperatingCashFlowRatio, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: cashFlowCoverageRatios, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: shortTermCoverageRatios, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: capitalExpenditureCoverageRatios, Length: 3808, dtype: bool
0       True
1       True
2       True
3       True
4       True
        ... 
3803    True
3804    True
3805    True
3806    True
3807    True
Name: dividendpaidAndCapexCoverageRati

In [13]:
df_14.dropna()

Empty DataFrame
Columns: [Unnamed: 0, Revenue, Revenue Growth, Cost of Revenue, Gross Profit, R&D Expenses, SG&A Expense, Operating Expenses, Operating Income, Interest Expense, Earnings before Tax, Income Tax Expense, Net Income - Non-Controlling int, Net Income - Discontinued ops, Net Income, Preferred Dividends, Net Income Com, EPS, EPS Diluted, Weighted Average Shs Out, Weighted Average Shs Out (Dil), Dividend per Share, Gross Margin, EBITDA Margin, EBIT Margin, Profit Margin, Free Cash Flow margin, EBITDA, EBIT, Consolidated Income, Earnings Before Tax Margin, Net Profit Margin, Cash and cash equivalents, Short-term investments, Cash and short-term investments, Receivables, Inventories, Total current assets, Property, Plant & Equipment Net, Goodwill and Intangible Assets, Long-term investments, Tax assets, Total non-current assets, Total assets, Payables, Short-term debt, Total current liabilities, Long-term debt, Total debt, Deferred revenue, Tax Liabilities, Deposit Liabilities, Total non-current liabilities, Total liabilities, Other comprehensive income, Retained earnings (deficit), Total shareholders equity, Investments, Net Debt, Other Assets, Other Liabilities, Depreciation & Amortization, Stock-based compensation, Operating Cash Flow, Capital Expenditure, Acquisitions and disposals, Investment purchases and sales, Investing Cash flow, Issuance (repayment) of debt, Issuance (buybacks) of shares, Dividend payments, Financing Cash Flow, Effect of forex changes on cash, Net cash flow / Change in cash, Free Cash Flow, Net Cash/Marketcap, priceBookValueRatio, priceToBookRatio, priceToSalesRatio, priceEarningsRatio, priceToFreeCashFlowsRatio, priceToOperatingCashFlowsRatio, priceCashFlowRatio, priceEarningsToGrowthRatio, priceSalesRatio, dividendYield, enterpriseValueMultiple, priceFairValue, ebitperRevenue, ebtperEBIT, niperEBT, grossProfitMargin, operatingProfitMargin, pretaxProfitMargin, netProfitMargin, effectiveTaxRate, returnOnAssets, returnOnEquity, returnOnCapitalEmployed, nIperEBT, ...]
Index: []

[0 rows x 225 columns]

In [14]:
for col in df_14.columns:
    dummy = df_14[df_14[col].isnull()]

In [15]:
dummy.shape

(0, 225)

We can not replace missing values and we can not drop rows which have missing values.

 Imputing the null with 0 because In our case missing values refects companies don't have value for that year.

In [16]:
df_14.fillna(0,inplace=True)
df_15.fillna(0,inplace=True)
df_16.fillna(0,inplace=True)
df_17.fillna(0,inplace=True)
df_18.fillna(0,inplace=True)

We choose to observe data of General Electic(GE).

In [17]:
if df_18['Unnamed: 0'].str.contains('GE').any() and df_14['Unnamed: 0'].str.contains('GE').any():
    print('"GE" is here')

"GE" is here


In [18]:
df_14 = df_14[df_14['Unnamed: 0'] =='GE' ]
df_15 = df_15[df_15['Unnamed: 0'] =='GE' ]
df_16 = df_16[df_16['Unnamed: 0'] =='GE' ]
df_17 = df_17[df_17['Unnamed: 0'] =='GE' ]
df_18 = df_18[df_18['Unnamed: 0'] =='GE' ]

In [19]:
df = df_14.append([df_15,df_16,df_17,df_18])#concatenating all the datasets
df.fillna(0,inplace=True)
df.index = [2014,2015,2016,2017,2018]

In [20]:
def graph(x):
    #this function will take a column of the dataset in between apostrophes
    #and will return a histogram of data over years
    print("Mean of "+ x+ " for GE last 5 year is {:.2f} ".format(df[x].mean()))
    plt.figure(figsize=(15, 7))
    df[x].plot.bar(color='y')
    plt.xlabel('Years')
    plt.ylabel(x)
    plt.title(x+ ' analysis GE ')
    plt.grid(False)
    plt.show()

In [21]:
graphs('Gross Profit')

NameError: name 'graphs' is not defined

In [ ]:
graphs('Revenue Growth')

In [ ]:
df.corr()

In [ ]:
corr = df.corr()
ax = sns.heatmap(
    corr, 
    vmin=-1, vmax=1, center=0,
    cmap=sns.diverging_palette(20, 220, n=200),
    square=True
)
ax.set_xticklabels(
    ax.get_xticklabels(),
    rotation=45,
    horizontalalignment='right'
);

In [ ]:
f = plt.figure(figsize=(19, 15))
plt.matshow(df.corr(), fignum=f.number)
plt.xticks(range(df.shape[1]), df.columns, fontsize=14, rotation=45)
plt.yticks(range(df.shape[1]), df.columns, fontsize=14)
cb = plt.colorbar()
cb.ax.tick_params(labelsize=14)
plt.title('Correlation Matrix', fontsize=16)

There are 225 columns in the dataset and to apply necessary algorithms we have to sort out the important columns which is to be done in future. This analysis will continue. Thanks for your kind attention.